### PR TITLE
Simple working PID Controller

### PR DESCRIPTION
**Pull request by frk2**
A working throttle/brake switching PID controller. In the absence of real twist commands, it was currently tested with the 'setpoint_node' which simply keeps alternating speed regularly. The autotuner sucks so the parameters were basically tuned by hand. An adaptive PID controller would be most suited for this task (maybe layer!)

The PID Controller is taken from: http://wiki.ros.org/pid (https://bitbucket.org/AndyZe/pid.git)

To debug, uncomment the following lines in dbw_sim.launch:
```
<node name="setpoint_node" pkg="pid" type="setpoint_node" output="screen" />
<node name="rqtplot" pkg="rqt_plot" type="rqt_plot" />
```

When you made, make sure you have qtpygraph installed or rqtplot would suck all the cores in the world and make your simulator lag.

Note 1: I have removed twist_controller.py and moved some of that code to dbw_node. Since our PID controller is a nice node, all dbw_node is doing is dispatching and processing inbound messages (which it'll do anyways) so adding another layer in the middle just seemed overkill. Let me know if you guys think otherwise.

Note 2: Also I could have added this as a submodule but I have subtle changes to the code to remove some dynamically configurable parameters that they use which were being very weird. So to save time I've just added this to the project. Otherwise I'd have to either figure out why the dynamic parametres are being weird OR fork it to my personal repo and then submodule it - seems like too much work for little gain

~~Note 3: The current_velocity sensor input is extremely noisy and sometimes quite way off (possible bug in the sim) hence our PID control is noisy - There is a low pass filter thats active but maybe use some sort of a 'average' or mean of the last inputs to make it smoother.~~

~~Note 4: Sometimes the simulator doesn't move, unknown why but seems like a common bug. Ignoring~~

Note 5:3/4 above is fixed by not repeatedly publishing to the simulator and increasing the queue size a bit!

### CODE REVIEW

**Comment by J-Rojas** [comment]
Not sure if this qualifies as 'external', but there's this note in the project guidelines:

"The project that is evaluated on the vehicle should not have any external libraries that aren't included in the starter repo."

I'll have to inquire Udacity about this before merging this in.

File: ros/src/twist_controller/dbw_node.py [hunk 1]
```diff
@@ -32,66 +39,120 @@
...
+        throttle = data.data
+        tcmd = ThrottleCmd()
+        tcmd.pedal_cmd_type = ThrottleCmd.CMD_PERCENT
+
+        bcmd = BrakeCmd()
+        #torque = Mass * acc * wheel_radius
+        max_brake_torque = min(BrakeCmd.TORQUE_MAX, (self.vehicle_mass + self.fuel_capacity * GAS_DENSITY)
+                               * abs(self.decel_limit) * self.wheel_radius)
+
+        if throttle >= 0:
```
Comment (J-Rojas) [inline ros/src/twist_controller/dbw_node.py]: If you're using a deadband, then why is the throttle engaged if greater than 0? Shouldn't it only be engaged if greater than the deadband?

File: ros/src/twist_controller/dbw_node.py [hunk 2]
```diff
@@ -32,66 +39,120 @@
...
         self.loop()
 
+    def on_control_effort(self, data):
+        throttle = data.data
+        tcmd = ThrottleCmd()
+        tcmd.pedal_cmd_type = ThrottleCmd.CMD_PERCENT
+
+        bcmd = BrakeCmd()
+        #torque = Mass * acc * wheel_radius
+        max_brake_torque = min(BrakeCmd.TORQUE_MAX, (self.vehicle_mass + self.fuel_capacity * GAS_DENSITY)
```
Comment (J-Rojas) [inline ros/src/twist_controller/dbw_node.py:95]: How did you determine this equation?

Comment (frk2) [reply ros/src/twist_controller/dbw_node.py:95]: Good question: http://sciencing.com/calculate-brake-torque-6076252.html

**Comment by J-Rojas** [comment]
Closing since there's another branch with these changes.

**Pull request closed without merging**